transmogrifier/helpers.py

```diff
@@ -83,6 +85,86 @@ def parse_xml_records(
             element.clear()
 
 
+def format_date(
+    date_value: Optional[str],
+) -> Optional[datetime]:
+    """
+    Format a date value as a datetime object according to one of the configured
+    OpenSearch date formats.
+
+    Args:
+        date_value: A date value.
+        source_record_id: The ID of the record being transformed.
+    """
+    if date_value:
+        for date_format in DATE_FORMATS:
+            try:
+                return datetime.strptime(date_value, date_format)
+            except ValueError:
+                pass
+    return None
+
+
+def validate_date(
+    date_value: Optional[str],
+    source_record_id: str,
+) -> Optional[str]:
+    """
+    Validate that a date can be parsed according to one of the configured
+    OpenSearch date formats. Returns original date value if valid or returns
+    None and logs an error.
+
+    Args:
+        date_value: A date value.
+        source_record_id: The ID of the record being transformed.
+    """
+    if date_value:
+        date_value = date_value.strip()
+        try:
+            if format_date(date_value):
+                return date_value
+        except ValueError:
+            pass
+    logger.error(
+        "Record # %s has a date that couldn't be parsed: %s",
+        source_record_id,
+        date_value,
+    )
+    return None
+
+
+def validate_date_range(
+    gte_date: Optional[str],
+    lte_date: Optional[str],
+    source_record_id: str,
+) -> bool:
+    """
+    Validate a date range by ensuring that the start date is before the end date to avoid
+    an OpenSearch exception. Returns true if only one date exists in the range or the end
+    date is after the start date, otherwise returns False and logs an error.
+
+    Args:
+        gte_date: The start date of a date range.
+        lte_date: The end date of a date range.
+        source_record_id: The ID of the record being transformed.
+    """
+    formatted_gte_date = format_date(gte_date)
```

Comment:
This should have two separate checks with two separate logging statements: one to validate that both values are valid date formats, and one to confirm that the end date is equal to or greater than the start date. Although the date validation function will also log an error in the first case, it won't be obvious that the issue was in a date range as opposed to a single date unless we also log it here.